packages/squiggle-lang/src/rescript/Utility/E/E_A.res

```diff
@@ -323,32 +323,45 @@ module Floats = {
       let continuous: array<float> = []
       let discrete = FloatFloatMap.empty()
 
-      let addData = (count: int, value: float): unit => {
-        if count >= minDiscreteWeight {
-          FloatFloatMap.add(value, count->Belt.Int.toFloat, discrete)
+      // In a run of exactly minDiscreteWeight, the first and last
+      // element indices differ by minDistance.
+      let minDistance = minDiscreteWeight - 1
+
+      let len = length(sortedArray)
+      let i = ref(0)
+      while i.contents < len - minDistance {
+        let value = sortedArray[i.contents]
+        if value != sortedArray[i.contents + minDistance] {
+          Js.Array2.push(continuous, value)->ignore
+          i := i.contents + 1
         } else {
-          for _ in 1 to count {
-            continuous->Js.Array2.push(value)->ignore
+          // Move i forward to next unequal value
+          // That is, find j so that equalAt(j-1) and !equalAt(j)
+          let i0 = i.contents
+          // Find base so that j is in (i0+base, i0+2*base]
+          let base = ref(minDistance)
+          let equalAt = (ind: int) => ind < len && sortedArray[ind] == value
+          while equalAt(i0 + base.contents * 2) {
```

Comment:
If much of this is very much like a binary search, have you thought about trying to pull that part out into a more generic binary search function?

Reply:
Yes, and there's a binary search utility I could use (although it would require taking a slice, which I think would have to copy and defeat the point). There are enough fiddly variations on binary search that it's easier to judge correctness when it's written out. Does the search function get the last equal index or the first unequal one? Does it read from the first and last indices or not?

Reply:
> Yes, and there's a binary search utility I could use (although it would require taking a slice, which I think would have to copy and defeat the point). 

First, I'm not suggesting you use an existing utility. It would be fine to write our own. 

Second, are you sure any function we write would involve taking a slice? I would have expected we could just pass in the indexes to start/stop from.

Reply:
I agree with Marshall. Binary search is surprisingly hard to abstract away in a library.
(e.g. every time I have to use [lower_bound](https://en.cppreference.com/w/cpp/algorithm/lower_bound) and [upper_bound](https://en.cppreference.com/w/cpp/algorithm/upper_bound) in C++, I have to double-check whether I'm doing the right thing and handling all corner cases, and it's not because STL is badly designed there.)